### ISSUE TITLE
Question: why chromium binary and shared libraries are compressed and what actually happens in a cold start?

### ISSUE THREAD

**Issue by germanocorrea** [alixaxel/chrome-aws-lambda#87]
I don't quite understand what happens in lambda cold starts, so I wonder why chromium and libraries are compressed in this package. Does lambda download the packages in cold start? Or they are compressed only for uploading purposes?

I've noticed that using 230MB for a PDF creation function (yes, I know I should allocate more, but I was just testing it) cold start takes about 20 seconds to execute, while the subsequent executions takes actually something like 5. I have wondered what would happen if I tried to use that stuff uncompressed. Would the function's cold start take more time to execute?

I'm asking here since I didn't find the AWS official documentation very helpful in this subject, it doesn't describe what actually happen in a cold start (not that I've found at least).

**Comment by extraordinaire** [comment]
@germanocorrea on a cold start the files get decompressed into `/tmp` (the decompression will take some time depending on what resources you allocate). On a hot start `/tmp` gets reused, so no need to decompress.

https://aws.amazon.com/blogs/compute/container-reuse-in-lambda/

The decompression [is avoided here](https://github.com/alixaxel/chrome-aws-lambda/blob/10feb8d162626d34aad2ee1e657f20956f53fe11/source/index.js#L151): if `/tmp/chromium` exists this is not a cold start so no need to decompress again.

**Comment by germanocorrea** [comment]
@extraordinaire I'm aware of that. I think I didn't made myself very clear, what I'm asking for is if the layer (or function) size interferes in the cold start time. Because if it doesn't, I could upload a layer where the binaries and libraries aren't compressed, right?

**Comment by germanocorrea** [comment]
**Update**: I've managed to use the binaries and libraries decompressed, had to also create some font configurations and zip some font files too (don't know why, I didn't investigated though) and apparently execution time don't change too much. So I guess compression is for uploading and to avoid the use of GitLarge File Storage?

**Comment by alixaxel** [comment]
@germanocorrea There are many limiting factors at play, but I'll to answer as best as I can.

## Git LFS

Yes, avoiding this is important - not for Lambda itself but for GCF & Firebase, where it isn't available.

## Lambda 50MB Limit

Unless you use Lambda Layers or upload through S3, we are limited to a 50MB package size.

You could unpack Chromium and provision it as a Layer, but you'd _still have to copy_ it to `/tmp`.

## Compressed vs Uncompressed

This is really hard to answer as Lambda is sort of a black box, but there are multiple factors at play.

The thing to keep in mind is that **all** resources in Lambda scale up as you allocate more memory.

### Compressed

1. Read 35MB (Read IO Bound)
2. Inflate (CPU & Memory Bound)
3. Write 130MB (Write IO Bound)

### Uncompressed

1. Read 130MB (Read IO Bound)
2. Write 130MB (Write IO Bound)

---

From my trials, there's a negligible benefit (100 to 200 ms) from unpacking Chromium.

Basically **whatever you save on CPU time you pay on disk read** operations. This is also the reason why I recommend using a 1600MB Lambda, as that seems to be the ideal sweet spot not only for warm-up times and Chromium performance but will also increase the function idle time (before it gets recycled).

Another optimization you could use (I do) is to send one (or multiple) hearbeat event every 10 minutes or so. You'll be charged for 100ms but you'll be able to keep X functions always in a warmed up state.
